docs/TensorLayout.md

```diff
@@ -0,0 +1,169 @@
+## Tensor Layout
```

Comment:
@shajrawi Thank you for the detailed doc!  I have a free questions. Could you give an example for the tensor layout and explain the semantics? Is the description optional (if we drop the layout hints will the program still be correct)? What does the optimizer need to know about the layout? Thank you.

Reply:
I have expanded the `Tensor Layout` section of the document to answer these questions. I also added more references to the related work section and expanded the future work section.